src/Emogrifier.php

```diff
@@ -670,42 +675,60 @@ private function parseCssRules($css)
     {
         $cssKey = md5($css);
         if (!isset($this->caches[static::CACHE_KEY_CSS][$cssKey])) {
-            // process the CSS file for selectors and definitions
-            preg_match_all('/(?:^|[\\s^{}]*)([^{]+){([^}]*)}/mi', $css, $matches, PREG_SET_ORDER);
-
-            $cssRules = [];
-            /** @var string[][] $matches */
-            /** @var string[] $cssRule */
-            foreach ($matches as $key => $cssRule) {
-                $cssDeclaration = trim($cssRule[2]);
-                if ($cssDeclaration === '') {
-                    continue;
-                }
-
-                $selectors = explode(',', $cssRule[1]);
-                foreach ($selectors as $selector) {
-                    // don't process pseudo-elements and behavioral (dynamic) pseudo-classes;
-                    // only allow structural pseudo-classes
-                    $hasPseudoElement = strpos($selector, '::') !== false;
-                    $hasAnyPseudoClass = (bool)preg_match('/:[a-zA-Z]/', $selector);
-                    $hasSupportedPseudoClass = (bool)preg_match(
-                        '/:(\\S+\\-(child|type\\()|not\\([[:ascii:]]*\\))/i',
-                        $selector
-                    );
-                    if ($hasPseudoElement || ($hasAnyPseudoClass && !$hasSupportedPseudoClass)) {
+            $splitCss = $this->splitCssAndMediaQuery($css);
+
+            $cssRules = [
+                'inlineable' => [],
+                'uninlineable' => []
+            ];
+            // keep track of where each rule appears in the file, since order is important
+            $key = 0;
+            foreach ($splitCss as $cssPart) {
+                // process each part for selectors and definitions
+                preg_match_all('/(?:^|[\\s^{}]*)([^{]+){([^}]*)}/mi', $cssPart['css'], $matches, PREG_SET_ORDER);
+
+                /** @var string[][] $matches */
+                /** @var string[] $cssRule */
+                foreach ($matches as $cssRule) {
+                    $cssDeclaration = trim($cssRule[2]);
+                    if ($cssDeclaration === '') {
                         continue;
                     }
 
-                    $cssRules[] = [
-                        'selector' => trim($selector),
-                        'declarationsBlock' => $cssDeclaration,
-                        // keep track of where it appears in the file, since order is important
-                        'line' => $key,
-                    ];
+                    $selectors = explode(',', $cssRule[1]);
+                    foreach ($selectors as $selector) {
+                        // don't process pseudo-elements and behavioral (dynamic) pseudo-classes;
```

Comment:
Can we have this part (within the foreach loop of $selector, or most of it) in a separate private method?

Reply:
I've moved the outer loop and regex matching into a separate method, so the loops are now very similar to how they were prior to this PR.